04-vue-cli/03-UiDropdown1/components/UiDropdown.vue

```diff
@@ -1,31 +1,106 @@
-    <button type="button" class="dropdown__toggle dropdown__toggle_icon">
-      <UiIcon icon="tv" class="dropdown__icon" />
-      <span>Title</span>
+  <div class="dropdown" :class="{ dropdown_opened: isOpened }">
+    <button
+      type="button"
+      class="dropdown__toggle"
+      :class="{ dropdown__toggle_icon: hasIcon }"
+      @click="
+        () => {
+          isOpened = !isOpened;
+        }
+      "
+    >
+      <UiIcon v-if="value?.icon" :icon="value.icon" class="dropdown__icon" />
+      <span>{{ value?.text || title }}</span>
     </button>
 
-    <div class="dropdown__menu" role="listbox">
-      <button class="dropdown__item dropdown__item_icon" role="option" type="button">
-        <UiIcon icon="tv" class="dropdown__icon" />
-        Option 1
-      </button>
-      <button class="dropdown__item dropdown__item_icon" role="option" type="button">
-        <UiIcon icon="tv" class="dropdown__icon" />
-        Option 2
+    <div v-show="isOpened" class="dropdown__menu" role="listbox">
+      <button
+        v-for="option in options"
+        class="dropdown__item"
+        :class="{ dropdown__item_icon: hasIcon }"
+        :key="option.value"
+        @click="() => select(option.value)"
+        role="option"
+        type="button"
+      >
+        <UiIcon v-if="option.icon" :icon="option.icon" class="dropdown__icon" />
+        {{ option.text }}
       </button>
     </div>
+
+    <select v-show="false" :value="value?.value" :text="value?.text" @change="selectNative">
+      <option v-for="option in options" :value="option.value" :key="option.value">{{ option.text }}</option>
+    </select>
   </div>
 </template>
 
-<script>
+<script lang="ts">
+import { defineComponent, type PropType } from 'vue';
 import UiIcon from './UiIcon.vue';
 
-export default {
+type TItem = { value: string; text: string; icon?: string };
+type TData = {
+  isOpened: boolean;
+};
+
+export default defineComponent({
   name: 'UiDropdown',
 
   components: { UiIcon },
-};
+
+  props: {
+    options: {
+      type: Array as PropType<TItem[]>,
+      default: () => [],
+    },
+
+    modelValue: {
+      type: String,
+    },
+
+    title: {
+      type: String,
+      required: true,
+    },
+  },
+
+  emits: ['update:modelValue'],
+
+  data(): TData {
+    return {
+      isOpened: false,
+    };
+  },
+
+  computed: {
+    hasIcon(): boolean {
+      return this.options.some(({ icon }) => icon);
+    },
+
+    value: {
+      get(): TItem | undefined {
+        return this.options.find(({ value }) => value === this.modelValue);
+      },
+
+      set({ value }: TItem) {
+        this.$emit('update:modelValue', value);
+      },
+    },
+  },
+
+  methods: {
+    select(value: string) {
+      this.value = this.options.find((it) => it.value === value);
+      this.isOpened = false;
+    },
+
+    selectNative(evt: Event) {
+      const { value } = evt.currentTarget as HTMLSelectElement;
+      this.select(value);
+    },
+  },
```

Comment:
There is no need to have 2 methods. We may call a method with another value:
`@change="select($event.target.value)"`

Reply:
I tried, but I got some errors:
![image](https://user-images.githubusercontent.com/47382827/230334781-b655bae1-2fdb-4ffc-8aed-7c8368f5f9d7.png)
and 
![image](https://user-images.githubusercontent.com/47382827/230334926-1ca57a5f-3a10-440f-8133-341724f71741.png)

Reply:
I've solved)

Reply:
I've solved)